### ISSUE TITLE
Highlight changes on login after notification email

### ISSUE THREAD

**Issue by aisapatino** [aisapatino/sjfnw#700]
Per Sunny, 12/30/14
>I have heard the complaint about how the notification email doesn't tell you what changed and then logging in to look it isn't easy to tell what has changed. can we make it so any changes appear highlighted upon the next login?

- [x] Style the notifications area so it stands out more
- [x] Include details such as donor name in the gift notification email

Then discuss possible indicators next to donor names in home page list (If we do that, I'd say we make it a separate issue)

**Comment by aisapatino** [comment]
This should be helped somewhat by #716, but some kind of highlight/notification would still be useful

**Comment by tarhata** [comment]
![gift_notification](https://cloud.githubusercontent.com/assets/8212941/7189303/f19dffbe-e432-11e4-8e6f-09141c67f2e4.png)

**Comment by tarhata** [comment]
How do you normally test emails and notifications locally? I got the above screenshot by using the live app. I think what SJF wants is for this notification to pop up every time `Received` or `Matched received` is updated on the admin page:

![image](https://cloud.githubusercontent.com/assets/8212941/7189900/18c80342-e437-11e4-85c5-4168bf76df06.png)

Also, aren't cron jobs supposed to run according to [`cron.yaml`](https://github.com/aisapatino/sjfnw/blob/master/cron.yaml#L16)? In this instance, shouldn't I have gotten the email at 17:09?

**Comment by aisapatino** [comment]
Emails don't get sent locally, but they're logged in the console instead so you can see when an email would have been sent. Notifications are annoying to test, so I put in a check a while back so that they aren't cleared locally ([this line in `home` view](https://github.com/aisapatino/sjfnw/blob/master/sjfnw/fund/views.py#L87). (The normal behavior is that a membership's notifications are cleared once they've been displayed.

So in other words you should be able see notifications locally even after refreshing. You can clear them by editing the membership through your local admin site. For emails, you have to check the console output.  Of course tests for both are good ways too.

I would check with them about whether they want a new notification every time one of those is _changed_. That would be a bit more complicated because we'd need a different message - if something saw "gift of $50 received from A" and then "gift of $70 received from A" they'd probably think $120 was received, rather than that staff corrected it from 50 to 70. Right now gift notifications are set up to be a one time thing.There should be notifications for the first time a gift is received per donor - see [these tests](https://github.com/aisapatino/sjfnw/blob/master/sjfnw/fund/tests/test_cron.py)

Cron time is UTC, so about 17:09 would be 12:09pm for us right now I believe. Also, on local you can go directly to the URL if you want to trigger it early (I wouldn't do that on live because it might create unexpected notifications for other users). This one is at `/mail/gifts`

**Comment by tarhata** [comment]
17:09 UTC is [apparently 10:09 Seattle time](https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=17:09+utc), but I got a notification email at 12:18pm :confused:  Does it take a long time for the cron job to process?

I'm at the office right now and talked to Sunny about it and they'd like for members to be notified whenever they receive funds from the donor, so yeah we would need to make different messages to accommodate the scenario you laid out above. But what if donors gift in payments, how does the system currently track that?

Also, just for clarity: when you say "gift notifications" you're referring to both the email and message/image that pops up when members log in?

**Comment by tarhata** [comment]
So I changed the time for the cron job locally, but don't see any email output in the console or see the notification when I log in to `/fund`. If you have time to check this out, that'd be really helpful!

**Comment by tarhata** [comment]
So I hashed this out with Sunny some more and the real issue might be that members want to be able to somehow "track" (I use that word loosely) the latest activity of their donors. For example, it could be as simple as adding an asterisk to a donor's name when SJF receives money from them.

![contacts](https://cloud.githubusercontent.com/assets/8212941/7191746/a9f3fe7c-e444-11e4-9373-b3e99bdf15a9.png)

But if notification emails are only sent once, then is it important that they know this? There's already a progress chart of their progress to track the amount of money they've fundraised.

**Comment by aisapatino** [comment]
By notifications I mean the message at the top of the PC home page, but it could have been confusing because they're tied to the emails in the code - as emails are compiled, notifications are set, too.

Did you also trigger the cron job by going to `/mail/gifts` in your browser? Try that and then check the console. Oh and it won't work more than once for the same donor unless you reset the donor's `gift_notified` to false.

Not sure how they do it if somebody pays in installments. Staff may enter it all at once, or update it. That's not handled in project central right now.

So it sounds like what they want is
1) Track updates to gift/match received, not just the initial entry
2) In a user's donor list, display some kind of indicator on donors for whom a gift/match was recently received.

Questions
- Should people get an **email** every time a gift/match is updated, or just the first time? (but show notifications/indicators every time)
  - I recommend only one email, otherwise there's a risk of email spam, for example if you typo the amount received and fix it, the person could get two emails in a row.
- If we're tracking updates, there are more complicated cases, how do we handle..
  - Negative amounts? (I.e. the received amount is reduced)
  - Multiple changes on the same donor? (Do we compile them, which would be tricky to do, or show all of them, which could be weird if amounts are being adjusted)
- If we do indicators by each donor, how long should they stay showing? Or do they just show once and disappear on reload like the notifications. (Showing once is the easiest to implement)

Technically
- We can easily style the notifications area to stand out more, if we want to still use it. I like it because it's explicit, and I think having that and donor indicators would work fine.
- I'm thinking we add a `notification` field to `Donor` that we set whenever received/matched changes.
  - Then we replace `membership.notifications` with a method `get_notifications()` that compiles them from each of the donors
- To detect changes to donor received/matched fields, we'll probably need to customize `Donor`'s `save` method and check for changes on the relevant fields.
- We may need additional field(s) if we want to show donor indicators more than once.

(This is kind of rough brainstorming, so ask questions if something doesn't make sense!)

**Comment by tarhata** [comment]
Nop, didn't know I had to trigger it by going to `/mail/gifts`. I see it all now! :)

For fields like `gift_notified` which aren't shown in `/admin`, do you normally manually change them using the Django python shell or mysql? For example, to set `gift_notified` to false:
```
mysql> UPDATE sjf_db 
    -> SET gift_notified=0 
    -> WHERE id=7549;
```

Sunny also suggested that maybe the email should explicitly say the names of donors who made a gift or pledge since the current email is too generic. Plus, seeing a person's name is much more memorable/personable and would probably cause people to go to PC and schedule a thank. 

I'm all for trying the easiest/simplest things first, so how about this:
1. Make the notification stand out more.  (do one-time)
2. Customize email message with donor's name
3. Put an indicator or something next to the donor's name in the `/fund` contacts list (do one-time)

If doing those things doesn't work out, then I'll probe one of the project managers or GP members to see what's really causing confusion and what would help. Don't think we should spend time implementing new fields and methods if the way we're thinking about this isn't in line with what people expect.

**Comment by aisapatino** [comment]
Weird, I left a comment yesterday (via email) that seems to have gone into the abyss. Mysql shell is probably the easiest; I haven't tried using the django shell for much. That command should do it :)

I agree about trying the simplest things first before doing something more complex. I would say go ahead with 1 & 2 and then we can talk about 3 because that will be a more complicated to set up.

**Comment by aisapatino** [comment]
Fixed by #755
